### ISSUE TITLE
Import cli.feature from cucumber-js

### ISSUE THREAD

**Issue by jbpros** [cucumber-attic/cucumber-tck#6]
[cli.feature](https://github.com/cucumber/cucumber-js/blob/master/features/cli.feature) now exists on cucumber-js.

First, it needs to be refactored into a more declarative style, like the other common features.

What are we expecting to be the common set of functionalities between implementations regarding the CLI? Should all implementation command-line interfaces respond to the same options and syntax? A subset of it? Should we spec the CLI at all (or leave its shape open to implementors)?

WDYT?

**Comment by cedric-lamalle** [comment]
I do think that there should be a set of common functionalities between the various cucumber implementations, and that these common functionalities should be using the same command line switches. Working with cucumber-jvm I've felt that options like '--dry-run' for instance are needed.

**Comment by jbpros** [comment]
Thanks for your comment, Cédric. And thank you for your great work on -jvm!

Disclaimer: I'm thinking out loud here :)

Now do you think we should arbitrarily select some of those functionalities or all of them? I think there are "core" CLI functionalities (like selecting features and scenarios, how the code for step definitions and support is loaded, etc.) that are definitely essential. Things like tags and dry runs are also important.

I was convinced a CLI feature is needed. But now I'm reflecting on it, I'm not so sure anymore.

First, do we want to enforce the CLI syntax? What would be the benefit of having all implementations follow the same syntax? What are the potential issues implementations could face because of that choice?

Another concern of mine is the risk of duplicating tests from other features. Take tags for example. We're going to write a feature to describe the expected behaviour of tags. Then, basically, in cli.feature we will duplicate those expectations (or a subset of them) with a focus on the exact syntax we want to include or exclude scenarios based on tags. I.e. we will keep the existing Given and Then steps and only change the Whens. But we are still going to describe the same behaviour: filtering scenarios based on tags.

This gets even worse if we decide not to enforce the syntax and go for a more declarative approach: the scenarios in both features are likely to be identical as the action steps (Whens) won't be different.

In short, I have two open questions:
- What would be the benefits of imposing the syntax to implementors?
- Aren't we going to only duplicate scenarios (as a more imperative form) from other features?

**Comment by mattwynne** [comment]
The benefits of imposting the syntax to implementors would be that users who switched from one platform to another would find the tool familiar to use. I think that's a good goal to have, at least for the _features_ that are common to all platforms.

That said, I wonder whether the right place to document these is not in one big fat 'CLI' feature but in lots of little features: tags_cli.feature, running_features_cli.feature, running_specific_scenarios_cli.feature etc.

**Comment by msassak** [comment]
I agree this is a good goal to have, but I'm not sure how valuable it is in relation to other functionality. @mattwynne can you comment on what you mean with "the _features_"? The italics are making me think I've missed something. 

For the time being I think we should kick this can down the road. So long as there's a simple way to execute a feature and see it fail, I don't think users switching from one to the other will be too overwhelmed with some syntax differences. Preserving the ambiguity of the CLI interface also encourages some experimentation with it, e.g. renaming dry-run to something more meaningful or splitting up its functionality into a set of flags.

**Comment by mattwynne** [comment]
I agree with you @msassak about diversity. My point in highlighting _features_ was expanded in my second paragraph. I don't think it's useful to specify the user interface to heavily anyway, we should specify each of the features which you happen to be able to use through the user interface.

**Comment by jbpros** [comment]
Good point Matt. A "familiar tool" would help people go from one implementation to another quite without confusion.

Moving CLI.feature scenarios to topic features is a neat idea. I guess duplicating some "non-cli" scenarios in them with an emphasis on the invokation syntax would be ok. 

And it leaves the door open for non-CLI implementations to run the suite (e.g. a GUI-only) as all "*_cli" features could still be excluded.

**Comment by jbpros** [comment]
In short:
- On one hand: define a closed interface to keep consistency and not confuse users; 
- On the other: leave the interface open so that implementors (or at least Mike :D) can experiment with new/other syntaxes.

Another solution is to _recommend_ the CLI syntax to implementors, without enforcing it in cucumber-features. This could be documented in some way (refer to `cucumber -h`?).

**Comment by msassak** [comment]
@mattwynne thanks for the clarification. That makes a lot of sense. @jbpros I think if we go with Matt's suggestion of features for what's available through the CLI we can sidestep the open/closed issue surrounding the syntax in favor of features defining a superset of CLI functionality available in one Cucumber or another, and the subset of what a particular implementation supports. So long as a CLI can "run a feature" or "list steps" or other common tasks, I suspect it won't be too hard to switch between the two, even taking into account differences in syntax.

**Comment by jbpros** [comment]
Fair enough. Let's leave the CLI syntax aside for now.
